Include raw FxA user-id and key-id in Hawk token.

rfk opened the pull request:
This is a sketch of how we might implement #121, sending the raw FxA uid and key-id as part of the hawk token id.   It will need some tests, and it may be an opportunity to refactor out some common helper functions for this stuff.

Connects to #121

/cc @bbangert @philbooth @pjenvey

philbooth commented on tokenserver/views.py:
```diff
@@ -151,29 +151,31 @@ def _valid_browserid_assertion(request, assertion):
-    request.metrics['fxa_uid'] = fxa_uid
+    hashed_fxa_uid = hashed_fxa_uid_full[:32]
+    request.validated['hashed_fxa_uid'] = hashed_fxa_uid
+    request.metrics['hashed_fxa_uid'] = hashed_fxa_uid
```
Just bike-shedding this property name for a moment (sorry), but I quite like how your explanatory comment calls this a hashed metrics uid, it just makes the distinction a bit clearer. Would `hashed_metrics_uid` be a more helpful name here?

philbooth replied:
(or even just `metrics_uid` as the hashing seems like an implementation detail)

rfk replied:
Bikeshedding very welcome, although we'll need to take care not to break any consumers of the existing names (e.g. changing the key used for `request.metrics` might not be possible without also updating some downstream lua filters).  I think including "metrics" in the name explicitly makes sense, esp. since the main reason we do it this way is for amplitude.

rfk commented:
OK, I've cleaned this up by extracting duplicated logic into the main `valid_authorization` function, and I've added some basic tests.  I think it's ready for an r? but **DO NOT MERGE** until we've sanity-checked whether any metrics scripts will break by renaming `request.metrics["fxa_uid"]` to `request.metrics["metrics_uid"]`.

rfk commented:
Ben, can you please review or delegate?

rfk commented:
> I assume the random key we'll want to use is the hashed fxa uid + the kid?

I recommend trying to somehow us a compound primary key, either as two columns or as a concatenated string "uid + kid", so that you can enumerate all data stored for a particular uid.

rfk commented:
(The `uid` will contain all the randomly-distributed-ness that you need for sharding etc)

bbangert commented:
Cool, Spanner lets us do composite keys, so we can include both of them as the primary key.

rfk commented:
> DO NOT MERGE until we've sanity-checked whether any metrics scripts will break by
> renaming request.metrics["fxa_uid"] to request.metrics["metrics_uid"].

Mailing list discussion didn't yield any blockers to this renaming; see https://bugzilla.mozilla.org/show_bug.cgi?id=1500830 for some followups if interested.

rfk commented:
> DO NOT MERGE until we've sanity-checked whether any metrics scripts will break by
> renaming request.metrics["fxa_uid"] to request.metrics["metrics_uid"].

Mailing list discussion didn't yield any blockers to this renaming; see https://bugzilla.mozilla.org/show_bug.cgi?id=1500830 for some followups if interested.

rfk commented:
I have a follow-up change I want to do to enable more accurate tracking of the "fxa_kid" property, but that depends on some FxA-side changes so I'm going to go ahead and merge this.

rfk commented:
Issue for the follow-up change: https://github.com/mozilla-services/tokenserver/issues/124
